### PR TITLE
Add FSDP config for Granite model

### PR DESCRIPTION
**Pull request by kplau1128**
Added FSDP config for Granite model

### CODE REVIEW

**Comment by kplau1128** [comment]
Test command line:
```
cd optimum-habana/examples/language-modeling

PT_HPU_AUTOCAST_LOWER_PRECISION_OPS_LIST=ops_bf16.txt PT_HPU_LAZY_MODE=0 \
python3 ../gaudi_spawn.py --world_size 8 --use_mpi run_lora_clm.py \
--model_name_or_path ibm-granite/granite-3.1-8b-instruct \
--dataset_name tatsu-lab/alpaca \
--bf16 True \
--output_dir ./result_ft_lora_granite-3.1-8b_bs1_4k_fsdp_8hpu \
--max_seq_len 4096 \
--per_device_train_batch_size 1 \
--per_device_eval_batch_size 1 \
--gradient_checkpointing \
--save_strategy no \
--learning_rate 0.0004 \
--warmup_ratio 0.03 \
--lr_scheduler_type "constant" \
--logging_steps 1 \
--dataset_concatenation \
--do_train \
--use_habana \
--throughput_warmup_steps 3 \
--lora_rank 4 \
--lora_target_modules "q_proj" "v_proj" "k_proj" "o_proj" \
--attn_softmax_bf16 True \
--validation_split_percentage 4 \
--use_lazy_mode False \
--fsdp_config granite_fsdp_config.json \
--fsdp auto_wrap \
--num_train_epochs 15 \
--eval_strategy epoch \
--eval_delay 2 \
--do_eval \
--pipelining_fwd_bwd False \
--use_fused_rope False \
--torch_compile_backend hpu_backend \
--torch_compile \
--use_flash_attention True \
--flash_attention_causal_mask True  2>&1 | tee result_ft_lora_granite-3.1-8b_bs1_4k_fsdp_8hpu.log
```

**Comment by imangohari1** [comment]
> Test command line:
> 
> ```
> cd optimum-habana/examples/language-modeling
> 
> PT_HPU_AUTOCAST_LOWER_PRECISION_OPS_LIST=ops_bf16.txt PT_HPU_LAZY_MODE=0 \
> python3 ../gaudi_spawn.py --world_size 8 --use_mpi run_lora_clm.py \
> --model_name_or_path ibm-granite/granite-3.1-8b-instruct \
> --dataset_name tatsu-lab/alpaca \
> --bf16 True \
> --output_dir ./result_ft_lora_granite-3.1-8b_bs1_4k_fsdp_8hpu \
> --max_seq_len 4096 \
> --per_device_train_batch_size 1 \
> --per_device_eval_batch_size 1 \
> --gradient_checkpointing \
> --save_strategy no \
> --learning_rate 0.0004 \
> --warmup_ratio 0.03 \
> --lr_scheduler_type "constant" \
> --logging_steps 1 \
> --dataset_concatenation \
> --do_train \
> --use_habana \
> --throughput_warmup_steps 3 \
> --lora_rank 4 \
> --lora_target_modules "q_proj" "v_proj" "k_proj" "o_proj" \
> --attn_softmax_bf16 True \
> --validation_split_percentage 4 \
> --use_lazy_mode False \
> --fsdp_config granite_fsdp_config.json \
> --fsdp auto_wrap \
> --num_train_epochs 15 \
> --eval_strategy epoch \
> --eval_delay 2 \
> --do_eval \
> --pipelining_fwd_bwd False \
> --use_fused_rope False \
> --torch_compile_backend hpu_backend \
> --torch_compile \
> --use_flash_attention True \
> --flash_attention_causal_mask True  2>&1 | tee result_ft_lora_granite-3.1-8b_bs1_4k_fsdp_8hpu.log
> ```

@kplau1128 what's the perf differences with and without using `granite_fsdp_config.json`?

**Comment by kplau1128** [comment]
Test result with `granite_fsdp_config.json` (~>5hrs):
``` json
{
    "epoch": 15.0,
    "eval_accuracy": 0.022623376514939612,
    "eval_graph_compliation_duration": 3.779285192489624,
    "eval_loss": 12.014018058776855,
    "eval_runtime": 13.245,
    "eval_samples": 59,
    "eval_samples_per_second": 3.701,
    "eval_steps_per_second": 0.529,
    "max_memory_allocated (GB)": 21.24,
    "memory_allocated (GB)": 4.29,
    "perplexity": 165052.36377976136,
    "total_flos": 5.4085885221666816e+17,
    "total_memory_available (GB)": 94.62,
    "train_loss": 1.808581720418529,
    "train_runtime": 20469.6554,
    "train_samples_per_second": 1.074,
    "train_steps_per_second": 0.135
}
```

Trying to run without `--fsdp_config `, it failing, it seems there a bug, need to debug it.

**Comment by kplau1128** [comment]
@imangohari1 this PR purpose is to enable FSDP for Granite out of the box as request.
The issue without `--fsdp / --fsdp_config` is existing bug need submit ticket to follow-up.

**Comment by imangohari1** [comment]
> @imangohari1 this PR purpose is to enable FSDP for Granite out of the box as request. The issue without `--fsdp / --fsdp_config` is existing bug need create separate ticket to follow-up.

Thanks for the update. My question is: how can we evaluate the perf/accuracy with or without fsdp on this model? what's the baseline we currently have for this model?

**Comment by kplau1128** [comment]
As I understanding, Granite3 model in OH has not yet optimization on Gaudi. It is use base class from transformers. Might need more work to optimum on Gaudi.
Related tickets:
https://habana.atlassian.net/browse/HS-5180
https://habana.atlassian.net/browse/HS-5623

**Review by imangohari1 — APPROVED** [review]
LGTM! 
The main change here is that the `"transformer_layer_cls_to_wrap": "GraniteDecoderLayer"` is needed for granite model. 

I tried running this on 8x A100-80g but I got OOM.

**Review by imangohari1 — APPROVED** [review]

**Comment by kplau1128** [comment]
@regisss could you please review, if it okay to merge?

**Comment by libinta** [comment]
@kplau1128  do you want to add a test case?

**Comment by kplau1128** [comment]
test case added:
`RUN_SLOW=1 python -m pytest ../../optimum-habana/tests/test_fsdp_examples.py -v -s -k granite-3.1-8b-instruct`
Tested on both Gaudi2/Gaudi3 passed

File: tests/baselines/fixture/tests/test_fsdp_examples.json [hunk 1]
```diff
@@ -18,5 +18,15 @@
...
     }
+  },
+  "tests/test_fsdp_examples.py::test_fsdp_bf16[ibm-granite/granite-3.1-8b-instruct--language-modeling-8-8-run_lora_clm.py-auto_wrap]": {
+    "gaudi2": {
+      "train_loss": 0.957,
+      "train_samples_per_second": 37.848
+    },
+    "gaudi3": {
+      "train_loss": 0.957,
+      "train_samples_per_second": 37.848
```
Comment (regisss) [inline tests/baselines/fixture/tests/test_fsdp_examples.json:29]: I get 32.009 on Gaudi3, which makes the test fail. Can you check please?

Comment (kplau1128) [reply tests/baselines/fixture/tests/test_fsdp_examples.json:29]: Interesting, last time when verify with `1.20.0-543`, I got ~43 on Gaudi3. Let me double check that.

Comment (emascarenhas) [reply tests/baselines/fixture/tests/test_fsdp_examples.json:29]: @kplau1128 , Do you have an update of the test results?

Comment (kplau1128) [reply tests/baselines/fixture/tests/test_fsdp_examples.json:29]: I not able to get Gaudi3 reservation to run, still waiting

Comment (regisss) [reply tests/baselines/fixture/tests/test_fsdp_examples.json:29]: 1.20, I don't have access to 1.21.
The thing is that there are many baselines that I can't reproduce on Gaudi3 so it's an issue that is not tightly related to this PR in particular. Let's see if the issue persists, going to merge this one.

**Review by regisss — APPROVED** [review]
